File: src/乐扣杯/3.py


class Solution:
    def minimumOperations(self, leaves: str) -> int:
        inf = float('inf')
        dp = [inf, inf, inf]

        for i, ch in enumerate(leaves):
            this = 0 if ch == 'r' else 1

            if this == 0:
                dp = [dp[0], min(dp[0], dp[1]) + 1, min(dp[1], dp[2])]
                if i == 0:
                    dp[0] = 0
            else:
                dp = [dp[0] + 1, min(dp[0], dp[1]), min(dp[1], dp[2]) + 1]
                if i == 0:
                    dp[0] = 1

        return dp[2]


if __name__ == '__main__':
    def test(input, expected):
        calc = Solution().minimumOperations(input)
        if calc != expected:
            print(f'case failed: `{input}`')
            print(f'  calc     = `{calc}`')
            print(f'  expected = `{expected}`')
            exit(1)
    test("rrryyyrryyyrr", 2)
    test("ryr", 0)
    test('rrrrrrrrrrrrrrr', 1)
    test('yyy', 2)
    test('yyyyyyyy', 2)
    test('ryyyy', 1)
    test('yyyyr', 1)
    test('rrr', 1)

File: src/乐扣杯/3.rs
struct Solution;

impl Solution {
    pub fn minimum_operations(leaves: String) -> i32 {
        const INF: i32 = i32::max_value() / 2;
        let mut dp = [INF, INF, INF];
        for (i, ch) in leaves.chars().enumerate() {
            if ch == 'r' {
                // 0
                dp = [dp[0], dp[0].min(dp[1]) + 1, dp[1].min(dp[2])];
                if i == 0 {
                    dp[0] = 0;
                }
            } else {
                // 1
                dp = [dp[0] + 1, dp[0].min(dp[1]), dp[1].min(dp[2]) + 1];
                if i == 0 {
                    dp[0] = 1;
                }
            }
        }

        dp[2]
    }
}

#[test]
fn test_solution() {
    macro_rules! test {
        ($args:tt, $ans:expr) => {
            assert_eq!(Solution::minimum_operations($args.into()), $ans)
        };
    }
    test!("rrryyyrryyyrr", 2);
    test!("ryr", 0);
    test!("rrrrrrrrrrrrrrr", 1);
    test!("yyy", 2);
    test!("yyyyyyyy", 2);
    test!("ryyyy", 1);
    test!("yyyyr", 1);
    test!("rrr", 1);
}
